apiserver/Makefile

```
@@ -56,11 +63,18 @@ imports: goimports ## Run goimports against code.
 	$(GOIMPORTS) -l -w .	
 
 test: fmt vet fumpt imports lint  ## Run unit tests.
-	go test ./... -race -coverprofile ray-kube-api-server-coverage.out
+	go test ./pkg/... ./cmd/... $(GO_TEST_FLAGS) -race -coverprofile ray-kube-api-server-coverage.out  -parallel 4
 
 lint: golangci-lint fmt vet fumpt imports ## Run the linter.
 	$(GOLANGCI_LINT) run  --timeout=3m	
 
+.PHONY: e2e-test
+e2e-test: ## Run end to end tests, a pre-exiting cluster.
+	go test ./test/e2e  $(GO_TEST_FLAGS) -timeout 30m  -race -parallel 4
+	
+.PHONY: local-e2e-test
+local-e2e-test: docker-image cluster load-image load-operator-image deploy-operator deploy e2e-test clean-cluster ## Run end to end tests, create a fresh kind cluster will all components deployed.
```

Should we pull/build a nightly operator image first before running `load-operator-image` ? 
```
(kuberay) root@cataract1:~/kuberay/apiserver# kind get clusters
No kind clusters found.
(kuberay) root@cataract1:~/kuberay/apiserver# make local-e2e-test
go fmt ./...
go vet ./...
test -s /root/kuberay/bin/gofumpt || GOBIN=/root/kuberay/bin go install mvdan.cc/gofumpt@v0.3.1
/root/kuberay/bin/gofumpt -l -w .
test -s /root/kuberay/bin/goimports || GOBIN=/root/kuberay/bin go install golang.org/x/tools/cmd/goimports@latest
/root/kuberay/bin/goimports -l -w .
test -s /root/kuberay/bin/golangci-lint || (curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | bash -s -- -b /root/kuberay/bin/ v1.54.1)
/root/kuberay/bin/golangci-lint run  --timeout=3m
go test ./pkg/... ./cmd/...  -race -coverprofile ray-kube-api-server-coverage.out  -parallel 4
?   	github.com/ray-project/kuberay/apiserver/pkg/client	[no test files]
?   	github.com/ray-project/kuberay/apiserver/pkg/interceptor	[no test files]
?   	github.com/ray-project/kuberay/apiserver/pkg/manager	[no test files]
ok  	github.com/ray-project/kuberay/apiserver/pkg/model	0.528s	coverage: 78.4% of statements
?   	github.com/ray-project/kuberay/apiserver/pkg/server	[no test files]
?   	github.com/ray-project/kuberay/apiserver/pkg/swagger	[no test files]
ok  	github.com/ray-project/kuberay/apiserver/pkg/util	0.311s	coverage: 51.2% of statements
?   	github.com/ray-project/kuberay/apiserver/cmd	[no test files]
docker build -t kuberay/apiserver:latest -f Dockerfile ..
[+] Building 40.2s (19/19) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                      0.0s
 => => transferring dockerfile: 32B                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                         0.0s
 => => transferring context: 2B                                                                                                                           0.0s
 => [internal] load metadata for registry.access.redhat.com/ubi8/go-toolset:1.19.10-10                                                                    0.0s
 => [stage-1 1/3] WORKDIR /workspace                                                                                                                      0.0s
 => [internal] load build context                                                                                                                         0.0s
 => => transferring context: 57.33kB                                                                                                                      0.0s
 => [builder  1/11] FROM registry.access.redhat.com/ubi8/go-toolset:1.19.10-10                                                                            0.0s
 => CACHED [builder  2/11] WORKDIR /workspace                                                                                                             0.0s
 => CACHED [builder  3/11] COPY apiserver/go.mod apiserver/go.mod                                                                                         0.0s
 => CACHED [builder  4/11] COPY apiserver/go.sum apiserver/go.sum                                                                                         0.0s
 => CACHED [builder  5/11] COPY proto/ proto/                                                                                                             0.0s
 => CACHED [builder  6/11] COPY ray-operator/ ray-operator/                                                                                               0.0s
 => [builder  7/11] COPY apiserver/ apiserver/                                                                                                            0.2s
 => [builder  8/11] WORKDIR /workspace/apiserver                                                                                                          0.1s
 => [builder  9/11] RUN go mod download                                                                                                                   9.1s
 => [builder 10/11] WORKDIR /workspace/apiserver                                                                                                          0.1s
 => [builder 11/11] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o kuberay-apiserver cmd/main.go                                               30.4s
 => CACHED [stage-1 2/3] COPY --from=builder /workspace/apiserver/kuberay-apiserver apiserver/                                                            0.0s
 => CACHED [stage-1 3/3] COPY proto/ proto/                                                                                                               0.0s
 => exporting to image                                                                                                                                    0.0s
 => => exporting layers                                                                                                                                   0.0s
 => => writing image sha256:e99e2d31e11094f0f550b26405884a76282a68f10a346bcb3d251ee23e0ac8ab                                                              0.0s
 => => naming to docker.io/kuberay/apiserver:latest                                                                                                       0.0s
/root/kuberay/bin/kind create cluster -n ray-api-server-cluster --config hack/kind-cluster-config.yaml
Creating cluster "ray-api-server-cluster" ...
 ✓ Ensuring node image (kindest/node:v1.23.17) 🖼
 ✓ Preparing nodes 📦 📦 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
 ✓ Joining worker nodes 🚜
Set kubectl context to "kind-ray-api-server-cluster"
You can now use your cluster with:

kubectl cluster-info --context kind-ray-api-server-cluster

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
/root/kuberay/bin/kind load docker-image kuberay/apiserver:latest -n ray-api-server-cluster
Image: "kuberay/apiserver:latest" with ID "sha256:e99e2d31e11094f0f550b26405884a76282a68f10a346bcb3d251ee23e0ac8ab" not yet present on node "ray-api-server-cluster-control-plane", loading...
Image: "kuberay/apiserver:latest" with ID "sha256:e99e2d31e11094f0f550b26405884a76282a68f10a346bcb3d251ee23e0ac8ab" not yet present on node "ray-api-server-cluster-worker2", loading...
Image: "kuberay/apiserver:latest" with ID "sha256:e99e2d31e11094f0f550b26405884a76282a68f10a346bcb3d251ee23e0ac8ab" not yet present on node "ray-api-server-cluster-worker", loading...
/root/kuberay/bin/kind load docker-image kuberay/operator:nightly -n ray-api-server-cluster
ERROR: image: "kuberay/operator:nightly" not present locally```

---

Without the building a local operator image the  load-operator-image would fail.
```suggestion
local-e2e-test: docker-image cluster load-image operator-image load-operator-image deploy-operator deploy e2e-test clean-cluster ## Run end to end tests, create a fresh kind cluster will all components deployed.
```

---

@z103cbI I ran local-e2e-test again this time it pulled but fail to start operator due to health prob timeout. Should we let people specify operator image for example, `make local-e2e-test -e OPERATOR_IMAGE_TAG=v1.0.0-rc.0`. Should we do `ifneq(${OPERATOR_IMAGE_TAG}, latest)`

---

@blublinksy what is your take on what version of the operator the apiserver end to end test should run on ?